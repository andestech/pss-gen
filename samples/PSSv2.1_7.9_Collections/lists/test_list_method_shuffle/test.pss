/*!
@weakgroup      test
@{
@file
@author         Luther Lee
@data           2023/10/12
@brief          Test list methods: `shuffle()`.
@version
Portable Test and Stimulus Standard Version 2.1 <br>
> Section 7.9 : Collections <br>
> Section 7.9.3 : Lists <br>


---------------------------------------------------------------------------------------------------

### Test Command
@code{.unparsed}
make gentarget
@endcode
*/
component pss_top {
    enum eSTR2NUM {
        ZERO, ONE, TWO
    };

    struct sSTR2NUM {
        string stringVal;
        int intVal;
    };

    action root_a {
        list<int     > intList    = {5      , 6      , 7   };
        list<bit [8] > byteList   = {8'd9   , 8'd8   , 8'd7};
        // list<bool    > boolList   = {true   , false  , true};   // todo: PSSListInst.assign(), Ln142: The list type should be assigned as a list_literal.
        list<bool    > boolList                             ;
        list<string  > stringList = {"Hello", "World", "!" };
        list<eSTR2NUM> enumList   = {TWO    , ONE    , TWO };
        // list<sSTR2NUM> structList = {"NULL" , 0            };   // todo: PSSInst.assign(), Ln321: PSSStructInst::assign is not impelemented
        list<sSTR2NUM> structList                           ;

        sSTR2NUM dummyStruct;

        int intVal_0 = 0;
        int intVal_1 = 0;
        int intVal_2 = 0;
        int intVal_3 = 0;
        int intVal_4 = 0;
        int intVal_5 = 0;

        exec pre_solve {
            dummyStruct.stringVal = "World"   ;
            dummyStruct.intVal    = 6         ;
            // structList.insert(0, dummyStruct );  // todo: PSSModel.getText(), Ln284: PSSRefModel::getText is not impelemented
        }

        exec post_solve {
            /** Test `shuffle()` */
            intList.shuffle()   ;
            byteList.shuffle()  ;
            boolList.shuffle()  ;
            stringList.shuffle();
            enumList.shuffle()  ;
            structList.shuffle();

            intVal_0 = intList.size()   ;
            intVal_1 = byteList.size()  ;
            intVal_2 = boolList.size()  ;
            intVal_3 = stringList.size();
            intVal_4 = enumList.size()  ;
            intVal_5 = structList.size();
        }

        exec body ASM =
"""
/** Test `shuffle()` */
intList [0:2] = {{intList[0]}}, {{intList[1]}}, {{intList[2]}}
byteList[0:2] = {{byteList[0]}}, {{byteList[1]}}, {{byteList[2]}}
stringList[0:2] = {{stringList[0]}}, {{stringList[1]}}, {{stringList[2]}}
enumList[0:2] = {{enumList[0]}}, {{enumList[1]}}, {{enumList[2]}}
""";
    }
}

/**@}*/
