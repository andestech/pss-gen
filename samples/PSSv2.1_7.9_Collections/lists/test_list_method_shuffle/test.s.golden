
/** Test `shuffle()` */
intList [0:2] = 6, 5, 7
byteList[0:2] = 8, 9, 7
stringList[0:2] = World, Hello, !
enumList[0:2] = 2, 2, 1

